### PR TITLE
wasm: Add minimum malloc size and split threshold

### PR DESCRIPTION
**Pull request by patrick-east**
To help reduce fragmentation we will now have a minimum of 16 byte
sized allocation block. In testing with JSON parsed objects we
generate a significant amount of 12 and 16 byte allocations for the
container structs for opa values, the size chosen should fit the
majority of them. In addition we will not split free blocks unless
the remainder is big enough to make additional allocations.

Testing with a relatively large (6MB) nested JSON object being parsed
this cuts the length of our free block list from a few thousand down
to zero. The time it takes to parse it (on my machine) goes from about
30 seconds to a few hundred milliseconds.

Signed-off-by: Patrick East <east.patrick@gmail.com>

### CODE REVIEW

**Comment by patrick-east** [comment]
😆  meant to make this a draft, still need to fix tests. Updates coming soon.

Was curious if anyone had thoughts on the minimum size and the threshold condition.

**Comment by patrick-east** [comment]
Should be OK now, although would definitely like suggestions for the malloc unit tests and resetting the memory state.

File: wasm/src/malloc.c [hunk 1]
```diff
@@ -98,16 +100,11 @@ void opa_heap_top_set(unsigned int top)
...
 {
-    init();
-    HEAP_CHECK();
-
-    // Look for the first free block that is large enough. Split the found block if necessary.
-
     for (struct heap_block *b = heap_free_start.next; b != &heap_free_end; b = b->next)
     {
-        if (b->size > (sizeof(struct heap_block) + size))
+        if (b->size > (sizeof(struct heap_block) + size + MIN_ALLOCATION_SZ))
```
Comment (tsandall) [inline wasm/src/malloc.c:107]: Is `MIN_ALLOCATION_SZ` intended to include the size of the heap_block struct? If not, do we need to add that (or just double `sizeof(struct heap_block)`)?

File: wasm/tests/test.c [hunk 2]
```diff
@@ -51,8 +60,84 @@ void test_opa_malloc()
...
+{
+    reset_heap();
+
+    // Ensure that allocations less than the minimum size
+    // are creating blocks large enough to be re-used by
+    // the minimum size.
+    void *too_small = opa_malloc(4);
+    test("allocated min block", too_small != NULL);
+
+    void *barrier = opa_malloc(0);
```
Comment (tsandall) [inline wasm/tests/test.c:73]: It wasn't immediately obvious to me why `barrier` is needed. Will the free blocks go to zero without `barrier`?

**Review by tsandall — COMMENTED** [review]
LGTM. Mostly just questions. Perhaps we can just live w/ the memory leak in the tests?

File: wasm/src/malloc.c (hunk 1 above)
Comment (koponen-styra) [reply wasm/src/malloc.c:107]: Both heap_blocks are already covered: the above sizeof(struct heap_block) is for the remaining block and b->size doesn't include the heap_block size.

File: wasm/tests/test.c (hunk 2 above)
Comment (koponen-styra) [reply wasm/tests/test.c:73]: right, the block would be removed from the free list without this.

**Review by koponen-styra — APPROVED** [review]
LGTM

File: wasm/src/malloc.c (hunk 1 above)
Comment (tsandall) [reply wasm/src/malloc.c:107]: ah, that makes sense now. thanks!

**Pull request merged**